### PR TITLE
feat(workflows): add reusable-build-artifact

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Summary

Add `reusable-build-artifact.yml` for Node build + artifact handoff (turbo-themes Build & Quality → Validate Examples; holy-grail Build & Test). Supports XOR `node-version` / JSON `node-version-matrix`, optional `post-build-test-command`, and static inner job names so GitHub appends matrix suffixes for org ruleset contexts.

## Type of Change

- [x] New feature (composite action, reusable workflow, or utility script)
- [ ] Bug fix
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD improvement

## Checklist

- [x] I have tested these changes locally
- [x] I have updated documentation if needed
- [x] Shell scripts pass `shellcheck`
- [x] YAML files pass `yamllint`
- [x] Python code passes `ruff` and `black`

## Breaking Changes

None

## Closes

- Closes #522

## Summary by CodeRabbit

* **New Features**
  * Added a reusable build workflow for Node.js projects.
  * Supports single-version or matrix builds, optional post-build tests, and artifact upload for downstream jobs.
  * Provides artifact details, including name, ID, and URL.
  * Adds validation for build inputs and prevents artifact naming conflicts across matrix builds.

* **Documentation**
  * Documented workflow configuration, permissions, outputs, handoff behavior, and merge queue compatibility.

* **Tests**
  * Added integration and unit coverage for workflow behavior, input validation, build execution, and artifact naming.

<details><summary><h3>Greptile Summary</h3></summary>

Adds `reusable-build-artifact.yml`, a new reusable workflow that runs a caller-supplied Node.js build command, an optional post-build test command, and uploads the resulting artifact for cross-job handoff. It supports both a single `node-version` and a JSON `node-version-matrix` (XOR), backed by three new shell scripts for input validation, artifact-name resolution, and build execution, plus a thorough bats test suite.

- The `prepare` → `build` matrix flow is well-structured: `validate-build-artifact-inputs.sh` enforces XOR version inputs and emits a CSV that `generate-version-matrix.sh` expands into the matrix; artifact-name collision across legs is correctly handled by appending `-<node-version>` in matrix mode.
- In the `prepare` job, `DEFAULT_VERSION` and `VERSIONS_INPUT` are both assigned `steps.validate.outputs.versions`; since the validate step guarantees a non-empty CSV, `VERSIONS_INPUT` is always consumed and `DEFAULT_VERSION` is dead code — diverging from the repo-wide pattern where `DEFAULT_VERSION` carries a meaningful singleton fallback.
- The test name `"rejects invalid JSON matrix"` misidentifies the rejection reason for `[20,22]` (valid JSON with integer elements), and the `resolve-build-artifact-name.sh` test suite is missing a case for the empty-`NODE_VERSION` rejection path that the script already guards against.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge; all shell scripts are well-guarded, the matrix/artifact logic is correct, and the bats suite provides solid end-to-end coverage of the new code paths.

The core build-and-upload pipeline is sound: XOR version validation, artifact-name deduplication across matrix legs, and artifact-path existence checks all behave correctly. The outstanding items are a cosmetic dead-code assignment (`DEFAULT_VERSION`), a misleading test name, and a missing test branch — none of which affect runtime behaviour.

`.github/workflows/reusable-build-artifact.yml` (dead `DEFAULT_VERSION` assignment) and `tests/bats/unit/actions/test_resolve_build_artifact_name.bats` (missing empty-`NODE_VERSION` test case).

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| .github/workflows/reusable-build-artifact.yml | New reusable workflow for Node build + artifact upload; `DEFAULT_VERSION` is dead code (always shadowed by `VERSIONS_INPUT`), and the initial `harden-runner` step omits `egress-preset`/`allowed-endpoints-mode` |
| scripts/ci/actions/validate-build-artifact-inputs.sh | Validates XOR version inputs, parses JSON matrix via inline Python; logic is correct and well-guarded |
| scripts/ci/actions/run-build-artifact.sh | Runs build and optional post-build-test commands in a subshell; validates artifact path exists afterward; correct use of `env -u BASH_ENV` |
| scripts/ci/actions/resolve-build-artifact-name.sh | Appends `-<node-version>` suffix in matrix mode; both empty-name and empty-version guards are present, but the unit-test suite only covers the empty-name branch |
| tests/bats/unit/actions/test_validate_build_artifact_inputs.bats | Good coverage of XOR validation and happy paths; the "rejects invalid JSON matrix" test name misidentifies the failure reason (`[20,22]` is valid JSON with integer elements, not malformed JSON) |
| tests/bats/unit/actions/test_resolve_build_artifact_name.bats | Covers single-version, matrix-mode, and empty-name cases; missing a test for the `NODE_VERSION must not be empty` rejection path |
| tests/bats/unit/actions/test_run_build_artifact.bats | Comprehensive coverage of build execution, post-build test, missing artifact, and missing working directory |
| tests/bats/integration/test_reusable_build_artifact.bats | Contract tests verify YAML structure (required fields, static job name, no pull-requests permission, pinned SHA); solid integration coverage |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Caller as Caller Workflow
    participant Prepare as prepare job
    participant Build as build job (matrix)
    participant Upstream as Downstream Jobs

    Caller->>Prepare: "node-version / node-version-matrix,<br/>build-command, artifact-name, artifact-path"
    Prepare->>Prepare: "validate-build-artifact-inputs.sh<br/>(XOR check, JSON parse)"
    Prepare->>Prepare: "generate-version-matrix.sh<br/>(emit JSON matrix)"
    Prepare-->>Build: "matrix={node-version:[…]}, matrix-mode"

    loop For each node-version in matrix
        Build->>Build: "setup-node @ matrix.node-version"
        Build->>Build: "resolve-build-artifact-name.sh<br/>(append -version if matrix-mode)"
        Build->>Build: "run-build-artifact.sh<br/>(build-command, post-build-test-command)"
        Build->>Build: "upload-artifact<br/>name=resolved, path=working-dir/artifact-path"
        Build-->>Upstream: artifact-name, artifact-id, artifact-url
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Caller as Caller Workflow
    participant Prepare as prepare job
    participant Build as build job (matrix)
    participant Upstream as Downstream Jobs

    Caller->>Prepare: "node-version / node-version-matrix,<br/>build-command, artifact-name, artifact-path"
    Prepare->>Prepare: "validate-build-artifact-inputs.sh<br/>(XOR check, JSON parse)"
    Prepare->>Prepare: "generate-version-matrix.sh<br/>(emit JSON matrix)"
    Prepare-->>Build: "matrix={node-version:[…]}, matrix-mode"

    loop For each node-version in matrix
        Build->>Build: "setup-node @ matrix.node-version"
        Build->>Build: "resolve-build-artifact-name.sh<br/>(append -version if matrix-mode)"
        Build->>Build: "run-build-artifact.sh<br/>(build-command, post-build-test-command)"
        Build->>Build: "upload-artifact<br/>name=resolved, path=working-dir/artifact-path"
        Build-->>Upstream: artifact-name, artifact-id, artifact-url
    end
```

</a>
</details>

<details><summary><h3>Comments Outside Diff (4)</h3></summary>

1. `.github/workflows/reusable-build-artifact.yml`, line 197-205 ([link](https://github.com/lgtm-hq/lgtm-ci/blob/099058dddba0d07b9149e669148d81984d7efba5/.github/workflows/reusable-build-artifact.yml#L197-L205)) 

   **`DEFAULT_VERSION` is dead code — deviates from the repo-wide pattern**

   Every other workflow that calls `generate-version-matrix.sh` sets `DEFAULT_VERSION` to a hardcoded single-version fallback (e.g. `${{ inputs.node-version || '20' }}` in `reusable-test-node.yml`) and `VERSIONS_INPUT` to the optional multi-version list. The script uses `VERSIONS_INPUT` when non-empty and falls back to `DEFAULT_VERSION` only when `VERSIONS_INPUT` is empty.

   Here, both are assigned the same expression (`steps.validate.outputs.versions`). Since the validate step guarantees a non-empty CSV, `VERSIONS_INPUT` is always populated and `DEFAULT_VERSION` is never consulted — making the `DEFAULT_VERSION` assignment dead code. The established pattern expects `DEFAULT_VERSION` to be a meaningful single-version fallback (e.g. `"20"`), so a future reader maintaining this workflow could be confused, and the pattern makes it harder to follow if `VERSIONS_INPUT` were ever relaxed to allow being empty.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

2. `tests/bats/unit/actions/test_resolve_build_artifact_name.bats`, line 45-54 ([link](https://github.com/lgtm-hq/lgtm-ci/blob/099058dddba0d07b9149e669148d81984d7efba5/tests/bats/unit/actions/test_resolve_build_artifact_name.bats#L45-L54)) 

   Missing test for the empty `NODE_VERSION` rejection path. The script errors with `NODE_VERSION must not be empty` when `NODE_VERSION` is blank, but there is no test covering that branch — unlike the analogous `ARTIFACT_NAME` empty test that is already present.

3. `tests/bats/unit/actions/test_validate_build_artifact_inputs.bats`, line 101-111 ([link](https://github.com/lgtm-hq/lgtm-ci/blob/099058dddba0d07b9149e669148d81984d7efba5/tests/bats/unit/actions/test_validate_build_artifact_inputs.bats#L101-L111)) 

   The test name says "invalid JSON" but `[20,22]` is perfectly valid JSON — it is a JSON array of integers. The actual rejection reason is that entries are not strings. Renaming the test makes it easier to understand why it is expected to fail.

   

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

4. `.github/workflows/reusable-build-artifact.yml`, line 258-261 ([link](https://github.com/lgtm-hq/lgtm-ci/blob/099058dddba0d07b9149e669148d81984d7efba5/.github/workflows/reusable-build-artifact.yml#L258-L261)) 

   **No `cache` parameter on the Node.js setup step**

   `actions/setup-node` supports a `cache` parameter (`npm`, `yarn`, or `pnpm`) that enables transparent dependency caching without requiring callers to manage it in their build commands. Without it, every matrix leg re-downloads packages from the registry on each run. Since the PR description explicitly places dependency installation in the caller's `build-command`, callers who do not roll their own caching will see slower builds. Consider adding an optional `node-cache` input (defaulting to empty/disabled) so callers can opt in to automatic package manager caching.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!
</details>

<sub>Reviews (4): Last reviewed commit: ["fix(ci): isolate user build commands fro..."](https://github.com/lgtm-hq/lgtm-ci/commit/ae2cbbfa295913ebb8deb6a8d215b8f55c1a266b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44947623)</sub>

### CODE REVIEW

File: .github/workflows/reusable-build-artifact.yml [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+          raw.githubusercontent.com:443
+          pipelines.actions.githubusercontent.com:443
+          github-releases.githubusercontent.com:443
+          bun.sh:443
+          registry.npmjs.org:443
+          pypi.org:443
+          files.pythonhosted.org:443
+          astral.sh:443
+          releases.astral.sh:443
+      allowed-endpoints-mode:
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply .github/workflows/reusable-build-artifact.yml:94]: Added nodejs.org:443 in 64a6a68.
